python/daqconf/apps/readout_gen.py

```diff
@@ -81,15 +81,20 @@ def get_readout_app(DRO_CONFIG=None,
 
     link_to_tp_sid_map = {}
     slr_link = {}
+
     for link in DRO_CONFIG.links:
         if SOFTWARE_TPG_ENABLED:
             link_to_tp_sid_map[link.dro_source_id] = SOURCEID_BROKER.get_next_source_id("Trigger")
             SOURCEID_BROKER.register_source_id("Trigger", link_to_tp_sid_map[link.dro_source_id], None)
-        elif FIRMWARE_TPG_ENABLED:
-            if not link.dro_slr in link_to_tp_sid_map.keys():
-                link_to_tp_sid_map[link.dro_slr] = SOURCEID_BROKER.get_next_source_id("Trigger")
-                SOURCEID_BROKER.register_source_id("Trigger", link_to_tp_sid_map[link.dro_slr], None)
-                slr_link[link.dro_slr] = link.dro_source_id
+    if FIRMWARE_TPG_ENABLED:
+        for fwsid in SOURCEID_BROKER.get_all_source_ids("FW_TPG"):
+            slr = SOURCEID_BROKER.sourceid_map["FW_TPG"][fwsid]
+            link_to_tp_sid_map[slr] = fwsid
+            slr_link[slr] = link_to_tp_sid_map[slr]
+            SOURCEID_BROKER.register_source_id("Trigger", link_to_tp_sid_map[slr], None)
```

Comment:
You have to be careful here to call `get_next_source_id`, to ensure you're not trying to re-use a SourceID that was registered elsewhere.

Reply:
so the problem I encountered with get_next_source_id is that is starts chekcing source IDs at zero, but the firmware TP source IDs start at a non-zero value. so when I use get_next_source_id it will just return zero. Is there a way to include some offset to get_next_source_id so it doesn't just start at zero?